packages/doenetml-worker-rust/lib-doenetml-core/src/dep_graph/graph.rs

```diff
@@ -0,0 +1,307 @@
+use std::{collections::HashMap, fmt::Debug, hash::Hash};
+
+/// An abstract graph. The speed of lookups is determined by the `Taggable` type,
+/// which is used to map `Node -> usize` for lookups in the internal data structures.
+///
+/// Example using `HashMap` as the `Taggable` type:
+/// ```rust
+/// # use doenetml_core::dep_graph::graph::Graph;
+/// use std::collections::HashMap;
+/// // Make a graph
+/// // a -> b -> c
+/// // a -> c
+/// let mut graph = Graph::<String, HashMap<_, _>>::new();
+/// graph.add_node("a".into());
+/// graph.add_node("b".into());
+/// graph.add_node("c".into());
+/// graph.add_edge("a".into(), "b".into());
+/// graph.add_edge("a".into(), "c".into());
+/// graph.add_edge("b".into(), "c".into());
+/// assert_eq!(graph.walk_descendants("a".into()).collect::<Vec<_>>(), vec!["b", "c"]);
+/// ```
+#[derive(Debug)]
+pub struct Graph<Node: Clone + Eq + Hash + Debug, Index: Taggable<Node, usize>> {
```

Comment:
Just for completeness, I'd either add a constructor where one can supply the initialized `Index` or just add the `Default` trait bound to `Index` here.

Reply:
Could you explain more?

Reply:
Is it possible to create a `Graph` where `Taggable` doesn't implement `Default`? I don't think so, because the fields are private and your constructors require that it implements `Default`. So, it just seemed strange to define a type (`Graph` where `Taggable` doesn't implement `Default`) that couldn't be constructed.

There's no practical purpose to my comment, and I don't really care whether you add the `Default` trait bound. It's more just a fun exercise in trying to reason with Rust trait bounds.